server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -946,7 +946,7 @@ protected void finalize(List<SnapshotFiles> snapshots, int fileListGeneration, M
             for (String blobName : blobs.keySet()) {
                 if (indexShardSnapshotsFormat.isTempBlobName(blobName) || blobName.startsWith(SNAPSHOT_INDEX_PREFIX)) {
                     try {
-                        blobContainer.deleteBlob(blobName);
+                        blobContainer.deleteBlobIgnoringIfNotExists(blobName);
```

Comment:
what about the comment below this:
> // We cannot delete index file - this is fatal, we cannot continue, otherwise we might end up with references to non-existing files with references to non-existing files

This sounds very dangerous, and I wonder if we need special precaution here. I wonder for example if we should first write the new index file before starting to delete old ones. It sounds to me as if finalization bears the risk of losing all index files if there is a process crash between deleting old ones and writing new one. We currently compensate for this in `buildBlobStoreIndexShardSnapshots` by doing a file listing and loading individual snapshots, but I find it odd to rely on such a fallback logic.

Reply:
Yes, I wanted to look at this too. I agree we should write the shard index file first, fail hard if something goes wrong, and then delete the files that are not needed anymore. We can still keep the individual file listing in `buildBlobStoreIndexShardSnapshots`.